Introduce multiple instances of connected-peers protocol.

shamil-gadelshin opened the pull request:
This PR follows https://github.com/subspace/subspace/pull/1654 introducing the connected-peers protocol and adding multiple instances of the same protocol to provide additional connectivity for farmers: maintaining connections not only with other farmers but with other peer types as well.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

shamil-gadelshin committed:
Introduce mulitple instances of connected-peers protocol.

nazar-pc reviewed (commented):
I certainly don't like the idea of having two protocols on the wire instead of one. This is an interesting hack, but the fact that local node cares about different kinds of peers is not relevant to the wire protocol at all and should not be exposed there.

shamil-gadelshin committed:
Remove protocol name from connected-peers protocol

shamil-gadelshin committed:
Fix rebase conflicts.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs:
```diff
@@ -303,12 +303,15 @@ pub(super) fn configure_dsn(
         // maintain permanent connections between farmers
-        connection_decision_handler: Arc::new(|peer_info| {
+        special_connection_decision_handler: Arc::new(|peer_info| {
             matches!(peer_info, PeerInfo::Farmer { .. })
```
Would be nice to have utility methods `PeerInfo::is_X()`

nazar-pc commented on crates/subspace-networking/src/bin/subspace-bootstrap-node/main.rs:
```diff
@@ -181,7 +181,7 @@ async fn main() -> anyhow::Result<()> {
                 max_pending_outgoing_connections: pending_out_peers,
                 // maintain permanent connections with any peer
-                connection_decision_handler: Arc::new(|_| true),
+                general_connection_decision_handler: Arc::new(|_| true),
```
We need to think carefully about this, I think bootstrap node is one of those cases where we want to return `false`, don't we?

shamil-gadelshin replied:
It's better to return false here, I agree. Basically, having set true here we occupy a good share of bootstrap node connections without a good reason.

shamil-gadelshin replied:
I'll disable both handlers explicitly.

nazar-pc commented on crates/subspace-networking/src/node_runner.rs:
```diff
@@ -113,8 +116,10 @@ where
-    /// Defines whether we maintain a persistent connection.
-    connection_decision_handler: ConnectionDecisionHandler,
+    /// Defines whether we maintain a persistent connection for common peers.
+    general_connection_decision_handler: ConnectionDecisionHandler,
```
`ConnectionDecisionHandler` name is not great, it is not clear what kinds of decisions it is making :confused: 

Maybe `ConnectedPeersHandler` or `KeepAliveHandler`?

nazar-pc commented on crates/subspace-networking/src/node_runner.rs:
```diff
@@ -272,8 +280,11 @@ where
+            }
+            SwarmEvent::Behaviour(Event::SpecialConnectedPeers(event)) => {
+                self.handle_connected_peers_event(Right(event)).await;
             }
```
Please don't use `Left` and `Right` just to reuse the method, it looks somewhat ugly. Either inline the code or create two utility methods.

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -55,21 +55,22 @@ const KADEMLIA_PROTOCOL: &[u8] = b"/subspace/kad/0.1.0";
 /// The default maximum established incoming connection number for the swarm.
-const SWARM_MAX_ESTABLISHED_OUTGOING_CONNECTIONS: u32 = 50;
+const SWARM_MAX_ESTABLISHED_OUTGOING_CONNECTIONS: u32 = 80;
 /// The default maximum pending incoming connection number for the swarm.
-const SWARM_MAX_PENDING_INCOMING_CONNECTIONS: u32 = 50;
+const SWARM_MAX_PENDING_INCOMING_CONNECTIONS: u32 = 80;
 /// The default maximum pending incoming connection number for the swarm.
-const SWARM_MAX_PENDING_OUTGOING_CONNECTIONS: u32 = 50;
+const SWARM_MAX_PENDING_OUTGOING_CONNECTIONS: u32 = 80;
 // The default maximum connection number to be maintained for the swarm.
-const SWARM_TARGET_CONNECTION_NUMBER: u32 = 50;
+const SWARM_TARGET_CONNECTION_NUMBER: u32 = 30;
```
What is the reason fro increasing defaults for total connections?

shamil-gadelshin replied:
Previously, we had connection expiration on our constant redialing of random peers for better connectivity. Now, we can have permanent connections to some random unused peers. I increased the limit to have the same connection number for all the tasks we have.

nazar-pc reviewed (commented):
None of the comments are blocking, but I am curious about changed limits and the default behavior of bootstrap node.

shamil-gadelshin committed:
Refactor connected-peers protocol.

- rename connected peers  handlers
- change bootstrap node default connected peers handler
- etc
